extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java

```diff
@@ -29,44 +32,34 @@
  */
 public final class HibernateEntityEnhancer implements BiFunction<String, ClassVisitor, ClassVisitor> {
 
-    private static final BytecodeProvider PROVIDER = new org.hibernate.bytecode.internal.bytebuddy.BytecodeProviderImpl(
+    private static final BytecodeProviderImpl PROVIDER = new org.hibernate.bytecode.internal.bytebuddy.BytecodeProviderImpl(
             ClassFileVersion.JAVA_V11);
 
+    //Choose this set to include Jakarta annotations, basic Java types such as String and Map, Hibernate annotations, and Panache supertypes:
+    private static final CoreTypePool CORE_POOL = new CoreTypePool("jakarta.", "java.", "org.hibernate.",
+            "io.quarkus.hibernate.");
```

Comment:
> > also now registering the Panache super types as "core types" so they'd also benefit from caching, if used.
> 
> Note to self: this implies that if we have live-reload tests in Panache which have entities using the same package name as Panache itself they might fail now for the silly reason of us skipping the work on such classes.

It also implies that demos using the `org.hibernate` package will no longer have live-reload... That's about every single demo the Hibernate team worked on. Certainly every single demo of mine.

Can't this list be generated at build time instead? So that we make extra sure not to include any application classes in the "reused" pool?

Reply:
> It also implies that demos using the `org.hibernate` package will no longer have live-reload... That's about every single demo the Hibernate team worked on. Certainly every single demo of mine.

It may also affect test case templates, at least if you port this stuff to Hibernate ORM...

Package-based checks really seem like they could bring a lot of trouble when working with the community (demos/bug reports) :/

Reply:
I agree and that's why the list is not hardcoded..  in Hibernate ORM I've refrained from adding the `org.hibernate` package to avoid such troubles.

 - https://github.com/hibernate/hibernate-orm/pull/8263/files#diff-ec2d0f0c23fe593565384ac5681d0a4873b6bc7d06f44157eb9d02b81c394dabR39

We can do the same here if you prefer..  I just don't see it very likely that people will be using the Panache namespace to create their own applications. Besides, the rule of "we should really not materialize this class" is for WildFly but I'm not sure if it would cause any trouble for Quarkus, I'm avoiding to materialize application classes out of excess of caution but I don't think it matters in Quarkus specifically - which is part of why I've added this.

But I'd prefer to be cautious indeed, if you're still concerned after that ^ I'm happy to change it.

Reply:
> I'm not sure if it would cause any trouble for Quarkus, I'm avoiding to materialize application classes out of excess of caution but I don't think it matters in Quarkus specifically

It probably matters for our own test suite, given the various metaspace (classloader) leaks we still have in `@QuarkusUnitTest`. In the real world, I don't know for sure, but we've had classloader leaks on live-reload in the past, so I wouldn't be surprised if we still had some.

And of course there are probably other concerns than just memory leaks, and I know even less about these.

> We can do the same here if you prefer..

I would. I'd argue that even caching the `jakarta` package could mean trouble in some cases, but I'll grant you that's going a bit far :)

In the case of Quarkus we know precisely which classes are application classes, which is I was suggesting to build this knowledge in Quarkus' own core pool.

But now that I think about it, I wonder if Quarkus' own knowledge is really necessary. The persistence unit descriptors do already include a list of model package/classes, don't they? And IIRC we absolutely need all model classes to be in there, or there could be malfunctions in other areas unrelated to bytecode enhancement. In the case of Quarkus, I think that classes that are not listed there aren't even enhanced.
So... couldn't you rely on that and use it as a "denylist" instead of the current "allowlist"?

See:

https://github.com/quarkusio/quarkus/blob/ab26345d9f689b4785e27754eb088b19b2347bb5/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaModelBuildItem.java#L75-L80

https://github.com/quarkusio/quarkus/blob/469fd3ae2e670dc7f43224242bb234edc591caf1/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L887-L897

Reply:
This isn't just about the model classes. Let me explain: what I observed during profiling is that when an entity class is introspected by ByteBuddy as it's building its model, it also describes all symbols referred to it:
 - the `@Entity` annotation on the entity makes it look for the class of the Entity annotation
 - a property having "String" as type makes it look for the String class
 - a propery having a "Map" makes it look for Map..
 - any other interface, annotation, etc..  on the mapped type or any signature of its methods and fields.

That's a lot of loading and bytecode parsing for each entity we're enhancing, as (before this patch) they would be repeated for each entity being enhanced.
After lookups their bytearray is parsed and BB extracts the datastructure it needs from them - this is what I'm caching, to not have to re-parse the structure of common types such as `@Entity` , "@Id" thousands of times - and record each of them in internal maps taking memory and making lookups heavy. So primarily JPA annotations, but one can see there's plenty other interesting annotations e.g. `jakarta.annotations.Generated` is hit fairly common.

The observation is that nobody is going to *modify* the bytecode @Entity annotation - and if they did, in this particular case livereload would actually still work as what ByteBuddy needs here about such objects is just to make sure the exist and are valid to be placed in a certain place.

So I included the Panache super class, because (if using Panache) you're most likely going to need that symbol for each and every entity again.

In summary, the "deny" list wouldn't include these, making it a little more dangerous.

On the other hand, the cache refactoring is already quite effective even w/o adding "dangerous" types to the CoreTypePool - they are effective even adding them to the ModelTypePool : the difference is that Quarkus is holding the CoreTypePool cache across the whole lifecycle of the JVM in dev mode, while we create a fresh instance of `ModelTypePool` for each live-reload - so we're still benefitting from the cache of ALL types during enhancement of N entities, but we throw it away and restart on the next live-reload.

So I'd suggest: let's limit it to more precise package names only?  I'd be fine to not cache `jakarta.annotations.Generated` but do cache the `jakarta.persistence` and `org.hibernate.annotations`.

If people then have problems because their app is using these packages I believe they deserve it :)

Reply:
> In summary, the "deny" list wouldn't include these, making it a little more dangerous.

I'm not sure what "these" is referring to exactly, but if it's the Panache super classes, they actually should be included in the list of classes of each persistence unit.

Anyway, that's a moot point, because:

> So I'd suggest: let's limit it to more precise package names only? I'd be fine to not cache `jakarta.annotations.Generated` but do cache the `jakarta.persistence` and `org.hibernate.annotations`.
> 
> If people then have problems because their app is using these packages I believe they deserve it :)

:+1: seems safe enough

Reply:
I think I'll go with:
```
"jakarta.persistence.", "java.", "org.hibernate.annotations.",                                                                    "io.quarkus.hibernate.reactive.panache.", "io.quarkus.hibernate.orm.panache",  "org.hibernate.search.mapper.pojo.mapping.definition.annotation", "jakarta.validation.constraints"
```

Any more you'd recommend?

It doesn't have to be comprehensive, so balancing things. Also don't want to have to check a long list for each loaded type , if there is no high likelyhood for it to pay back with substantial benefits.

Reply:
> "org.hibernate.search.mapper.pojo.mapping.definition.annotation"

And I wish this was shorter :P

Reply:
Amen to this. We all make mistakes.